[🐛 BUG] Lambda functions not working when page is imported

### What went wrong? 🤔

Lambda functions are not working anymore when the page is being imported into the main.py

Nothing appears on the page even if I ahve other visual elements in my page.

![image](https://github.com/user-attachments/assets/ba239ba5-d75f-4850-a918-cf390ccb6eb0)


![image](https://github.com/user-attachments/assets/8143bf42-bd88-4f66-88fa-ae00e571e098)


```python
[2024-10-03 11:47:17.965][Taipy][ERROR] Exception on /taipy-jsx/TaiPy_root_page [GET]
Traceback (most recent call last):
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\flask\app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\flask\app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\flask_cors\extension.py", line 194, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\flask\app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\flask\app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\gui.py", line 2409, in __render_page
    context = page.render(self)
              ^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\_page.py", line 38, in render
    self._rendered_jsx = self._renderer.render(gui)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\page.py", line 63, in render
    return self._base_element._render(gui)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\_element.py", line 183, in _render
    return f"{el[0]}{self._render_children(gui)}</{el[1]}>"
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\_element.py", line 186, in _render_children
    return "\n".join([child._render(gui) for child in self._children])
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\_element.py", line 186, in <listcomp>
    return "\n".join([child._render(gui) for child in self._children])
                      ^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\_element.py", line 262, in _render
    self._evaluate_lambdas(gui)
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\builder\_element.py", line 73, in _evaluate_lambdas
    gui._bind_var_val(k, _getscopeattr(gui, expr))
                         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jacta\.conda\envs\develop\Lib\site-packages\taipy\gui\utils\_attributes.py", line 22, in _getscopeattr
    return getattr(gui._get_data_scope(), name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'types.SimpleNamespace' object has no attribute '__lambda_2569653502496'     
```

### Expected Behavior

It should work fine like before or like the "{}" syntax.

### Steps to Reproduce Issue

Create these files: main.py, page.py

main.py:

```python
from other_page import page
import taipy as tp

value = 10

tp.Gui(page).run()

```

other_page.py:

```python
import taipy.gui.builder as tgb

with tgb.Page() as page:
    tgb.text(lambda value: str(value))
```

Run the code.

### Version of Taipy

develop - 10/3/24

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.


### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

--- a/.gitignore
+++ b/.gitignore
@@ -90,7 +90,7 @@ Data
 
 # demo files
 demo-*
-demo_*/
+demo_*
 .airflow
 *.dags
 data_sources

--- a/taipy/gui/_renderers/builder.py
+++ b/taipy/gui/_renderers/builder.py
@@ -460,7 +460,10 @@ class _Builder:
             val_list = value if isinstance(value, list) else [value]
             for val in val_list:
                 ret = self.__gui._run_adapter(
-                    t.cast(t.Callable, adapter), val, adapter.__name__ if isroutine(adapter) else "adapter", id_only=True
+                    t.cast(t.Callable, adapter),
+                    val,
+                    adapter.__name__ if isroutine(adapter) else "adapter",
+                    id_only=True,
                 )  # type: ignore
                 if ret is not None:
                     ret_list.append(ret)

--- a/taipy/gui/_renderers/builder.py
+++ b/taipy/gui/_renderers/builder.py
@@ -428,9 +428,9 @@ class _Builder:
                 var_type = self.__gui._get_unique_type_adapter(type(elt).__name__)
             if adapter is None:
                 adapter = self.__gui._get_adapter_for_type(var_type)
-            elif var_type == str.__name__ and callable(adapter):
+            elif var_type == str.__name__ and isroutine(adapter):
                 var_type += (
-                    f"__lambda_{id(adapter)}"
+                    _get_lambda_id(t.cast(LambdaType, adapter))
                     if adapter.__name__ == "<lambda>"
                     else _get_expr_var_name(adapter.__name__)
                 )
@@ -438,19 +438,19 @@ class _Builder:
                 if adapter is None:
                     adapter = self.__gui._get_adapter_for_type(lov_name)
                 else:
-                    self.__gui._add_type_for_var(lov_name, var_type)
+                    self.__gui._add_type_for_var(lov_name, t.cast(str, var_type))
             if value_name := self.__hashes.get("value"):
                 if adapter is None:
                     adapter = self.__gui._get_adapter_for_type(value_name)
                 else:
-                    self.__gui._add_type_for_var(value_name, var_type)
+                    self.__gui._add_type_for_var(value_name, t.cast(str, var_type))
             if adapter is not None:
                 self.__gui._add_adapter_for_type(var_type, adapter)  # type: ignore
 
             if default_lov is not None and lov:
                 for elt in lov:
                     ret = self.__gui._run_adapter(
-                        t.cast(t.Callable, adapter), elt, adapter.__name__ if callable(adapter) else "adapter"
+                        t.cast(t.Callable, adapter), elt, adapter.__name__ if isroutine(adapter) else "adapter"
                     )  # type: ignore
                     if ret is not None:
                         default_lov.append(ret)
@@ -460,7 +460,7 @@ class _Builder:
             val_list = value if isinstance(value, list) else [value]
             for val in val_list:
                 ret = self.__gui._run_adapter(
-                    t.cast(t.Callable, adapter), val, adapter.__name__ if callable(adapter) else "adapter", id_only=True
+                    t.cast(t.Callable, adapter), val, adapter.__name__ if isroutine(adapter) else "adapter", id_only=True
                 )  # type: ignore
                 if ret is not None:
                     ret_list.append(ret)

--- a/taipy/gui/builder/_element.py
+++ b/taipy/gui/builder/_element.py
@@ -39,6 +39,7 @@ class _Element(ABC):
     _TAIPY_EMBEDDED_PREFIX = "_tp_embedded_"
     _EMBEDDED_PROPERTIES = ["decimator"]
     _TYPES: t.Dict[str, str] = {}
+    __LAMBDA_VALUE_IDX = 0
 
     def __new__(cls, *args, **kwargs):
         obj = super(_Element, cls).__new__(cls)
@@ -64,6 +65,12 @@ class _Element(ABC):
     def update(self, **kwargs):
         self._properties.update(kwargs)
         self.parse_properties()
+
+    @staticmethod
+    def __get_lambda_index():
+        _Element.__LAMBDA_VALUE_IDX += 1
+        _Element.__LAMBDA_VALUE_IDX %= 10000
+        return _Element.__LAMBDA_VALUE_IDX
 
     def _evaluate_lambdas(self, gui: Gui):
         for k, lmbd in self._lambdas.items():
@@ -129,7 +136,7 @@ class _Element(ABC):
             tree = _TransformVarToValue(self.__calling_frame, args + targets + _python_builtins).visit(lambda_fn)
             ast.fix_missing_locations(tree)
             lambda_text = ast.unparse(tree)
-            lambda_name = _get_lambda_id(value)
+            lambda_name = _get_lambda_id(value, index=(_Element.__get_lambda_index()))
             self._lambdas[lambda_name] = lambda_text
             return f'{{{lambda_name}({", ".join(args)})}}'
         except Exception as e:

--- a/taipy/gui/builder/_element.py
+++ b/taipy/gui/builder/_element.py
@@ -69,7 +69,7 @@ class _Element(ABC):
     @staticmethod
     def __get_lambda_index():
         _Element.__LAMBDA_VALUE_IDX += 1
-        _Element.__LAMBDA_VALUE_IDX %= 10000
+        _Element.__LAMBDA_VALUE_IDX %= 0xFFFFFFF0
         return _Element.__LAMBDA_VALUE_IDX
 
     def _evaluate_lambdas(self, gui: Gui):

--- a/taipy/gui/builder/_element.py
+++ b/taipy/gui/builder/_element.py
@@ -16,13 +16,12 @@ import copy
 import inspect
 import re
 import typing as t
-import uuid
 from abc import ABC, abstractmethod
 from collections.abc import Iterable
 from types import FrameType, FunctionType
 
 from .._warnings import _warn
-from ..utils import _getscopeattr
+from ..utils import _get_lambda_id, _getscopeattr
 from ._context_manager import _BuilderContextManager
 from ._factory import _BuilderFactory
 from ._utils import _LambdaByName, _python_builtins, _TransformVarToValue
@@ -37,7 +36,6 @@ class _Element(ABC):
     _ELEMENT_NAME = ""
     _DEFAULT_PROPERTY = ""
     __RE_INDEXED_PROPERTY = re.compile(r"^(.*?)__([\w\d]+)$")
-    _NEW_LAMBDA_NAME = "new_lambda"
     _TAIPY_EMBEDDED_PREFIX = "_tp_embedded_"
     _EMBEDDED_PROPERTIES = ["decimator"]
     _TYPES: t.Dict[str, str] = {}
@@ -100,8 +98,8 @@ class _Element(ABC):
             if key.startswith("on_") or self._is_callable(key):
                 return value if value.__name__.startswith("<") else value.__name__
             # Parse lambda function_is_callable
-            if (lambda_name := self.__parse_lambda_property(key, value)) is not None:
-                return lambda_name
+            if (lambda_call := self.__parse_lambda_property(key, value)) is not None:
+                return lambda_call
         # Embed value in the caller frame
         if not isinstance(value, str) and key in self._EMBEDDED_PROPERTIES:
             return self.__embed_object(value, is_expression=False)
@@ -131,7 +129,7 @@ class _Element(ABC):
             tree = _TransformVarToValue(self.__calling_frame, args + targets + _python_builtins).visit(lambda_fn)
             ast.fix_missing_locations(tree)
             lambda_text = ast.unparse(tree)
-            lambda_name = f"__lambda_{uuid.uuid4().hex}"
+            lambda_name = _get_lambda_id(value)
             self._lambdas[lambda_name] = lambda_text
             return f'{{{lambda_name}({", ".join(args)})}}'
         except Exception as e:

--- a/taipy/gui/builder/_utils.py
+++ b/taipy/gui/builder/_utils.py
@@ -57,18 +57,28 @@ class _LambdaByName(ast.NodeVisitor):
 
     def visit_Call(self, node):
         if getattr(node.func, "attr", None) == self.element_name:
-            if self.lambdas.get(_LambdaByName._DEFAULT_NAME, None) is None:
-                self.lambdas[_LambdaByName._DEFAULT_NAME] = next(
+            if self.lambdas.get(_LambdaByName._DEFAULT_NAME, None) is None and (
+                a_lambda := next(
                     (
                         arg
                         for arg in node.args
-                        if isinstance(arg, ast.Lambda) and self.lineno >= arg.lineno and self.lineno <= arg.end_lineno
+                        if isinstance(arg, ast.Lambda)
+                        and arg.lineno is not None
+                        and arg.end_lineno is not None
+                        and self.lineno >= arg.lineno
+                        and self.lineno <= arg.end_lineno
                     ),
                     None,
                 )
+            ):
+                self.lambdas[_LambdaByName._DEFAULT_NAME] = a_lambda
+
             for kwd in node.keywords:
                 if (
-                    isinstance(kwd.value, ast.Lambda)
+                    kwd.arg is not None
+                    and isinstance(kwd.value, ast.Lambda)
+                    and kwd.value.lineno is not None
+                    and kwd.value.end_lineno is not None
                     and self.lineno >= kwd.value.lineno
                     and self.lineno <= kwd.value.end_lineno
                 ):

--- a/taipy/gui/gui.py
+++ b/taipy/gui/gui.py
@@ -13,7 +13,6 @@ from __future__ import annotations
 
 import contextlib
 import importlib
-import inspect
 import json
 import math
 import os
@@ -25,6 +24,7 @@ import typing as t
 import warnings
 from importlib import metadata, util
 from importlib.util import find_spec
+from inspect import currentframe, getabsfile, ismethod, ismodule, isroutine
 from pathlib import Path
 from threading import Timer
 from types import FrameType, FunctionType, LambdaType, ModuleType, SimpleNamespace
@@ -81,6 +81,7 @@ from .utils import (
     _get_client_var_name,
     _get_css_var_value,
     _get_expr_var_name,
+    _get_lambda_id,
     _get_module_name_from_frame,
     _get_non_existent_file_path,
     _get_page_from_module,
@@ -311,7 +312,7 @@ class Gui:
                 own Flask application instance and use it to serve the pages.
         """
         # store suspected local containing frame
-        self.__frame = t.cast(FrameType, t.cast(FrameType, inspect.currentframe()).f_back)
+        self.__frame = t.cast(FrameType, t.cast(FrameType, currentframe()).f_back)
         self.__default_module_name = _get_module_name_from_frame(self.__frame)
         self._set_css_file(css_file)
 
@@ -824,7 +825,7 @@ class Gui:
         if callable(on_change_fn):
             try:
                 arg_count = on_change_fn.__code__.co_argcount
-                if arg_count > 0 and inspect.ismethod(on_change_fn):
+                if arg_count > 0 and ismethod(on_change_fn):
                     arg_count -= 1
                 args: t.List[t.Any] = [None for _ in range(arg_count)]
                 if arg_count > 0:
@@ -1509,7 +1510,7 @@ class Gui:
         if callable(action_function):
             try:
                 argcount = action_function.__code__.co_argcount
-                if argcount > 0 and inspect.ismethod(action_function):
+                if argcount > 0 and ismethod(action_function):
                     argcount -= 1
                 args = t.cast(list, [None for _ in range(argcount)])
                 if argcount > 0:
@@ -1532,7 +1533,7 @@ class Gui:
         cp_args = [] if args is None else args.copy()
         cp_args.insert(0, self.__get_state())
         argcount = user_function.__code__.co_argcount
-        if argcount > 0 and inspect.ismethod(user_function):
+        if argcount > 0 and ismethod(user_function):
             argcount -= 1
         if argcount > len(cp_args):
             cp_args += (argcount - len(cp_args)) * [None]
@@ -2073,7 +2074,7 @@ class Gui:
                 self.add_page(name=k, page=v)
         elif isinstance(folder_name := pages, str):
             if not hasattr(self, "_root_dir"):
-                self._root_dir = os.path.dirname(inspect.getabsfile(self.__frame))
+                self._root_dir = os.path.dirname(getabsfile(self.__frame))
             folder_path = folder_name if os.path.isabs(folder_name) else os.path.join(self._root_dir, folder_name)
             folder_name = os.path.basename(folder_path)
             if not os.path.isdir(folder_path):  # pragma: no cover
@@ -2224,9 +2225,9 @@ class Gui:
     def _download(
         self, content: t.Any, name: t.Optional[str] = "", on_action: t.Optional[t.Union[str, t.Callable]] = ""
     ):
-        if callable(on_action) and on_action.__name__:
+        if isroutine(on_action) and on_action.__name__:
             on_action_name = (
-                _get_expr_var_name(str(on_action.__code__))
+                _get_lambda_id(t.cast(LambdaType, on_action))
                 if on_action.__name__ == "<lambda>"
                 else _get_expr_var_name(on_action.__name__)
             )
@@ -2377,7 +2378,7 @@ class Gui:
             return
         try:
             arg_count = on_page_load_fn.__code__.co_argcount
-            if arg_count > 0 and inspect.ismethod(on_page_load_fn):
+            if arg_count > 0 and ismethod(on_page_load_fn):
                 arg_count -= 1
             args: t.List[t.Any] = [None for _ in range(arg_count)]
             if arg_count > 0:
@@ -2752,7 +2753,7 @@ class Gui:
 
         app_config = self._config.config
 
-        run_root_dir = os.path.dirname(inspect.getabsfile(self.__frame))
+        run_root_dir = os.path.dirname(getabsfile(self.__frame))
 
         # Register _root_dir for abs path
         if not hasattr(self, "_root_dir"):
@@ -2797,7 +2798,7 @@ class Gui:
         glob_ctx: t.Dict[str, t.Any] = {t.__name__: t for t in _TaipyBase.__subclasses__()}
         glob_ctx[Gui.__SELF_VAR] = self
         glob_ctx["state"] = self.__state
-        glob_ctx.update({k: v for k, v in locals_bind.items() if inspect.ismodule(v) or callable(v)})
+        glob_ctx.update({k: v for k, v in locals_bind.items() if ismodule(v) or callable(v)})
 
         # Call on_init on each library
         for name, libs in self.__extensions.items():

--- a/taipy/gui/utils/_evaluator.py
+++ b/taipy/gui/utils/_evaluator.py
@@ -263,7 +263,7 @@ class _Evaluator:
             _warn(f"Cannot evaluate expression '{not_encoded_expr if is_edge_case else expr_string}'", e)
             expr_evaluated = None
         if lambda_expr and callable(expr_evaluated):
-            expr_hash = _get_lambda_id(expr_evaluated)
+            expr_hash = _get_lambda_id(expr_evaluated, module_name)
         # save the expression if it needs to be re-evaluated
         return self.__save_expression(gui, expr, expr_hash, expr_evaluated, var_map, lambda_expr)
 

--- a/taipy/gui/utils/_evaluator.py
+++ b/taipy/gui/utils/_evaluator.py
@@ -263,7 +263,7 @@ class _Evaluator:
             _warn(f"Cannot evaluate expression '{not_encoded_expr if is_edge_case else expr_string}'", e)
             expr_evaluated = None
         if lambda_expr and callable(expr_evaluated):
-            expr_hash = _get_lambda_id(expr_evaluated, module_name)
+            expr_hash = _get_lambda_id(expr_evaluated, module=module_name)
         # save the expression if it needs to be re-evaluated
         return self.__save_expression(gui, expr, expr_hash, expr_evaluated, var_map, lambda_expr)
 

--- a/taipy/gui/utils/_lambda.py
+++ b/taipy/gui/utils/_lambda.py
@@ -9,8 +9,11 @@
 # an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 # specific language governing permissions and limitations under the License.
 
+import typing as t
 from types import LambdaType
 
+from ._variable_directory import _variable_encode
 
-def _get_lambda_id(lambda_fn: LambdaType):
-    return f"__lambda_{id(lambda_fn)}"
+
+def _get_lambda_id(lambda_fn: LambdaType, module: t.Optional[str] = None):
+    return _variable_encode(f"__lambda_{id(lambda_fn)}", lambda_fn.__module__ or module)

--- a/taipy/gui/utils/_lambda.py
+++ b/taipy/gui/utils/_lambda.py
@@ -15,5 +15,7 @@ from types import LambdaType
 from ._variable_directory import _variable_encode
 
 
-def _get_lambda_id(lambda_fn: LambdaType, module: t.Optional[str] = None):
-    return _variable_encode(f"__lambda_{id(lambda_fn)}", lambda_fn.__module__ or module)
+def _get_lambda_id(lambda_fn: LambdaType, module: t.Optional[str] = None, index: t.Optional[int] = None):
+    return _variable_encode(
+        f"__lambda_{id(lambda_fn)}{f'_{index}' if index is not None else ''}", lambda_fn.__module__ or module
+    )

